### PR TITLE
[cbc] Add login support

### LINKED ISSUE #16764: [cbc.ca] Error 402: Payment Required
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.06.18*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [X] I've **verified** and **I assure** that I'm running youtube-dl **2018.06.18**

### Before submitting an *issue* make sure you have:
- [X] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [X] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [X] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [X] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [X] Question
- [ ] Other

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

```
[debug] System config: []
[debug] User config: ['--all-subs', '--sub-format', 'srt', '--convert-subs', 'srt', '--embed-subs', '--merge-output-format', 'mkv', '--recode-video', 'mkv', '--format', 'bestvideo[width<=?1280]+bestaudio/best', '-o', 'D:/Video/Downloads/%(series)s/%(series)s - %(season_number)sx%(episode_number)02d - %(title)s [%(height)s].%(ext)s', '--prefer-ffmpeg', '--ffmpeg-location', 'D:\\Video\\FFMPEG\\bin', '--download-archive', 'D:/Video/Downloads/Archive.txt']
[debug] Custom config: []
[debug] Command-line args: ['https://watch.cbc.ca/media/heartland/season-9/episode-7/38e815a-0094c1b411a', '--username', 'PRIVATE', '-v']
Type account password and press [Return]:
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2018.06.18
[debug] Python version 3.4.4 (CPython) - Windows-8.1-6.3.9600
[debug] exe versions: ffmpeg N-90771-g5079e96bcc, ffprobe N-90771-g5079e96bcc
[debug] Proxy map: {}
[debug] Using fake IP 99.239.129.198 (CA) as X-Forwarded-For.
[cbc.ca:watch] 38e815a-0094c1b411a: Downloading XML
[cbc.ca:watch] 38e815a-0094c1b411a: Downloading XML
[download] Downloading playlist: Fearless
[cbc.ca:watch] playlist Fearless: Collected 1 video ids (downloading 1 of them)
[download] Downloading video 1 of 1
[debug] Using fake IP 99.234.54.159 (CA) as X-Forwarded-For.
[cbc.ca:watch:video] c78f815d-d3ed-4502-871f-da9347e0318a: Downloading XML
ERROR: Unable to download XML: HTTP Error 402: Payment Required (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpu1zifx01\build\youtube_dl\extractor\common.py", line 579, in _request_webpage
  File "C:\Users\dst\AppData\Roaming\Build archive\youtube-dl\rg3\tmpu1zifx01\build\youtube_dl\YoutubeDL.py", line 2211, in urlopen
  File "C:\Python\Python34\lib\urllib\request.py", line 470, in open
  File "C:\Python\Python34\lib\urllib\request.py", line 580, in http_response
  File "C:\Python\Python34\lib\urllib\request.py", line 508, in error
  File "C:\Python\Python34\lib\urllib\request.py", line 442, in _call_chain
  File "C:\Python\Python34\lib\urllib\request.py", line 588, in http_error_default
```

---

### If the purpose of this *issue* is a *site support request* please provide all kinds of example URLs support for which should be included (replace following example URLs by **yours**):
- Single video: https://watch.cbc.ca/media/heartland/season-9/episode-7/38e815a-0094c1b411a

Note that **youtube-dl does not support sites dedicated to [copyright infringement](https://github.com/rg3/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free)**. In order for site support request to be accepted all provided example URLs should not violate any copyrights.

---

### Description of your *issue*, suggested solution and other information

Unlike Error 402 for Youtube which typically signifies too many requests, a restricted IP, or a CAPTCHA request, in this situation some of the videos on the site (mainly those not currently being broadcast but instead from a previous season, etc) require a free membership in order to watch them. In this situation, "Error 402: Payment Required" actually seems appropriate in that you need to be logged in to access select files. After trying multiple different options and variables, it appears that the youtube-dl interface is either not passing the login info to the page, or when logged in within the browser, the login info is not being save as a cookie that can be utilised by youtube-dl. Files that don't require logging in to view are easily downloaded.
Is there some sort of variable/option that I am missing that would circumvent the issue, or is it just not passing the credentials on to the site at the appropriate time?

### PR DESCRIPTION
**Pull request by meunierd**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Improvement

---

### Description of your *pull request* and other information

`_API_KEY` is defined in a JS file loaded by watch.cbc.ca, it's a static value that identifies the app. You are logging into this app via login radius when you POST your account credentials.

Requests are made by a device, which registers itself and uses the provided id and token. The default token is unauthenticated but you can log the device in with the access token you receive as a response from the account login. This provides a new authenticated device token, which will let you download gated videos!

Closes #16764

### CODE REVIEW

**Comment by Triforcecwp** [comment]
Thanks so much for working on this @meunierd , I really appreciate it.

**Comment by MrN00BZ** [comment]
Can I download CBC videos with this then or it is just in testing and if so how?

**Comment by meunierd** [comment]
You can checkout the branch and use your login credentials, it works for me

**Comment by MrN00BZ** [comment]
Sorry, but what does that mean how do you use it?

**Comment by meunierd** [comment]
If you have difficulty following the [developer instructions] (https://github.com/rg3/youtube-dl/blob/master/README.md#developer-instructions) then it's probably best to wait until this makes its way into a release

**Comment by meunierd** [comment]
@dstftw is this good to merge?

**Comment by smcclay** [comment]
I have not been able to get this to work. It's possible that I'm doing something wrong, but I have checked out this pull request and this is what I get when I run "python -m youtube_dl ..." :

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'-u', u'PRIVATE', u'-p', u'PRIVATE', u'https://watch.cbc.ca/media/kims-convenience/season-1/episode-1/38e815a-00ade4014dc']
[debug] Encodings: locale UTF-8, fs utf-8, out None, pref UTF-8
[debug] youtube-dl version 2019.01.30.1
[debug] Git HEAD: 7d569a37e
[debug] Python version 2.7.15 (CPython) - Darwin-18.2.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 4.1, ffprobe 4.1
[debug] Proxy map: {}
[debug] Using fake IP 99.238.54.128 (CA) as X-Forwarded-For.
[cbc.ca:watch] 38e815a-00ade4014dc: Downloading XML
[cbc.ca:watch] 38e815a-00ade4014dc: Downloading XML
[download] Downloading playlist: Gay Discount
[cbc.ca:watch] playlist Gay Discount: Collected 1 video ids (downloading 1 of them)
[download] Downloading video 1 of 1
[debug] Using fake IP 99.239.194.78 (CA) as X-Forwarded-For.
[cbc.ca:watch:video] ce247368-121a-4174-8b28-1879d2550f18: Downloading XML
ERROR: Unable to download XML: HTTP Error 402: Payment Required (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "youtube_dl/extractor/common.py", line 605, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "youtube_dl/YoutubeDL.py", line 2221, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python@2/2.7.15_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

**Comment by meunierd** [comment]
You might have an unauthenticated token in your cache, in which case it would use the cached one.

On macOS I can clear the cache by running `rm ~/.cache/youtube-dl/cbcwatch/device.json`

**Comment by smcclay** [comment]
That fixed it for me. Thanks for doing this.

**Comment by meunierd** [comment]
@dstftw how can I move this PR forward?

**Comment by dstftw** [comment]
You must provide account credentials.

File: youtube_dl/extractor/cbc.py [hunk 1]
```diff
@@ -249,13 +280,19 @@ def _valid_device_token(self):
...
+        signature = self._signature()
         self._device_id = self._device_token = None
         result = self._download_xml(
             self._API_BASE_URL + 'device/register',
             None, 'Acquiring device token',
             data=b'<device><type>web</type></device>')
         self._device_id = xpath_text(result, 'deviceId', fatal=True)
         self._device_token = xpath_text(result, 'deviceToken', fatal=True)
+        if signature:
+            data = '<login><token>{}</token><device><deviceId>{}</deviceId><type>web</type></device></login>'.format(signature, self._device_id).encode()
```
Comment (dstftw) [inline youtube_dl/extractor/cbc.py]: `{}` won't work in python 2.6.

Comment (meunierd) [reply youtube_dl/extractor/cbc.py]: Thanks! I've updated this to use `{0}` and `{1}`

**Comment by meunierd** [comment]
username: `hfkhum+ek9u6dei4bof8@sharklasers.com`
password: `password`

**Comment by meunierd** [comment]
There we go, forgot to push the commit, sorry about that!

**Comment by makeworld-the-better-one** [comment]
What happened to this, can it still be merged? Also @meunierd have you confirmed this will still work with gem.cbc.ca shows?

CC: @dstftw @remitamine

**Comment by meunierd** [comment]
Yeah I'm still using my branch. The last time I did was maybe 3 weeks ago,
I pushed a rebased copy up to work with the gem.cbc.ca switch. I posted
credentials, not sure what else I can do on my part.

**Comment by makeworld-the-better-one** [comment]
```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['https://gem.cbc.ca/media/kims-convenience/season-2/episode-1/38e815a-00d01e9a6e8', '-u', 'PRIVATE', '-p', 'PRIVATE', '-v']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2019.06.21
[debug] Python version 3.7.3 (CPython) - Linux-5.1.12-arch1-1-ARCH-x86_64-with-arch
[debug] exe versions: ffmpeg 4.1.3, ffprobe 4.1.3, rtmpdump 2.4
[debug] Proxy map: {}
[debug] Using fake IP 99.254.14.117 (CA) as X-Forwarded-For.
[cbc.ca:watch] 38e815a-00d01e9a6e8: Downloading XML
[cbc.ca:watch] 38e815a-00d01e9a6e8: Downloading XML
[download] Downloading playlist: Janet's Roommate
[cbc.ca:watch] playlist Janet's Roommate: Collected 1 video ids (downloading 1 of them)
[download] Downloading video 1 of 1
[debug] Using fake IP 99.244.104.206 (CA) as X-Forwarded-For.
[cbc.ca:watch:video] 35735f55-771d-4413-add1-dacc98555036: Downloading XML
ERROR: Unable to download XML: HTTP Error 402: Payment Required (caused by <HTTPError 402: 'Payment Required'>); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/lib/python3.7/site-packages/youtube_dl/extractor/common.py", line 627, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/lib/python3.7/site-packages/youtube_dl/YoutubeDL.py", line 2227, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
```

Doesn't seem to work for me.

**Comment by meunierd** [comment]
Just pushed up a rebased branch. Works for me, try clearing your ~/.cache

**Comment by meunierd** [comment]
Could it be that you need to proxy through Canada?

```
python -m youtube_dl 'https://gem.cbc.ca/media/kims-convenience/season-2/episode-1/38e815a-00d01e9a6e8'                                                           ~/s/youtube-dl:cbc-login<>
[cbc.ca:watch] 38e815a-00d01e9a6e8: Downloading XML
[cbc.ca:watch] 38e815a-00d01e9a6e8: Downloading XML
[download] Downloading playlist: Janet's Roommate
[cbc.ca:watch] playlist Janet's Roommate: Collected 1 video ids (downloading 1 of them)
[download] Downloading video 1 of 1
[cbc.ca:watch:video] 35735f55-771d-4413-add1-dacc98555036: Downloading XML
[cbc.ca:watch:video] 35735f55-771d-4413-add1-dacc98555036: Downloading XML
[cbc.ca:watch:video] 35735f55-771d-4413-add1-dacc98555036: Downloading m3u8 information
[download] Destination: Janet's Roommate-38e815a-00d0f131e87.f7093.mp4
ffmpeg version 4.1.3 Copyright (c) 2000-2019 the FFmpeg developers
  built with Apple LLVM version 10.0.1 (clang-1001.0.46.4)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/4.1.3_1 --enable-shared --enable-pthreads --enable-version3 --enable-hardcoded-tables --enable-avresample --cc=clang --host-cflags='-I/Library/Java/JavaVirtualMachines/adoptopenjdk-11.0.2.jdk/Contents/Home/include -I/Library/Java/JavaVirtualMachines/adoptopenjdk-11.0.2.jdk/Contents/Home/include/darwin' --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libbluray --enable-libmp3lame --enable-libopus --enable-librubberband --enable-libsnappy --enable-libtesseract --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libx264 --enable-libx265 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librtmp --enable-libspeex --enable-videotoolbox --disable-libjack --disable-indev=jack --enable-libaom --enable-libsoxr
  libavutil      56. 22.100 / 56. 22.100
  libavcodec     58. 35.100 / 58. 35.100
  libavformat    58. 20.100 / 58. 20.100
  libavdevice    58.  5.100 / 58.  5.100
  libavfilter     7. 40.101 /  7. 40.101
  libavresample   4.  0.  0 /  4.  0.  0
  libswscale      5.  3.100 /  5.  3.100
  libswresample   3.  3.100 /  3.  3.100
  libpostproc    55.  3.100 / 55.  3.100
[hls,applehttp @ 0x7ff89c800000] Opening 'https://cbcvott-a.akamaihd.net/p/38e815a-00d01e9a6e8/CBC_KIMS_CONVENIENCE_SEASON_02_S02E01-v3-14042644/segments/keys/cl-CBC-Crypt-64.key' for reading
[hls,applehttp @ 0x7ff89c800000] Opening 'crypto+https://cbcvott-a.akamaihd.net/p/38e815a-00d01e9a6e8/CBC_KIMS_CONVENIENCE_SEASON_02_S02E01-v3-14042644/segments/CBC_KIMS_CONVENIENCE_SEASON_02_S02E01_v9/v9_0.ts' for reading
Input #0, hls,applehttp, from 'https://cbcvott-a.akamaihd.net/p/38e815a-00d01e9a6e8/CBC_KIMS_CONVENIENCE_SEASON_02_S02E01-v3-14042644/segments/CBC_KIMS_CONVENIENCE_SEASON_02_S02E01_v9/prog_index.m3u8':
  Duration: 00:21:53.25, start: 10.033367, bitrate: 0 kb/s
  Program 0
    Metadata:
      variant_bitrate : 0
    Stream #0:0: Video: h264 (High) ([27][0][0][0] / 0x001B), yuv420p, 1920x1080 [SAR 1:1 DAR 16:9], Closed Captions, 29.97 fps, 29.97 tbr, 90k tbn, 59.94 tbc
    Metadata:
      variant_bitrate : 0
Output #0, mp4, to 'file:Janet's Roommate-38e815a-00d0f131e87.f7093.mp4.part':
  Metadata:
    encoder         : Lavf58.20.100
    Stream #0:0: Video: h264 (High) (avc1 / 0x31637661), yuv420p, 1920x1080 [SAR 1:1 DAR 16:9], q=2-31, 29.97 fps, 29.97 tbr, 90k tbn, 90k tbc
    Metadata:
      variant_bitrate : 0
```

**Comment by makeworld-the-better-one** [comment]
I did clear my cache, I'll try from the rebase later. I am in Canada right now, so I don't think a proxy is needed.

**Comment by makeworld-the-better-one** [comment]
I've confirmed it works.

**Comment by monty124** [comment]
no update on getting this merged?

**Comment by monty124** [comment]
Hi All, I can confirm this is working! would be great to merge.

**Comment by meunierd** [comment]
@dstftw I've addressed all comments, can I get a re-review?

**Comment by zeegbi** [comment]
@meunierd thank you for the commit, I've just tested it works like a charm!

File: youtube_dl/extractor/cbc.py [hunk 2]
```diff
@@ -216,6 +216,37 @@ class CBCWatchBaseIE(InfoExtractor):
...
+            'email': email,
+            'password': password,
+        }).encode()
+        headers = {'content-type': 'application/json'}
+        resp = self._download_json(
+            self._LOGIN_URL,
+            None,
+            data=data,
+            headers=headers,
+            query={'apikey': self._API_KEY}
```
Comment (dstftw) [inline youtube_dl/extractor/cbc.py]: Remove superfluous verbosity.

File: youtube_dl/extractor/cbc.py [hunk 3]
```diff
@@ -216,6 +216,37 @@ class CBCWatchBaseIE(InfoExtractor):
...
+        )
+        access_token = resp['access_token']
+
+        # token
+        query = {
+            "access_token": access_token,
+            "apikey": self._API_KEY,
+            "jwtapp": "jwt",
```
Comment (dstftw) [inline youtube_dl/extractor/cbc.py]: Single quotes.

**Comment by ilike2burnthing** [comment]
Can confirm that this is working without issue on my end! Thanks @meunierd 

Hopefully it will get merged soon by @dstftw or someone else 👍

**Comment by meunierd** [comment]
reping @dstftw this is ready for review

**Comment by bojangles89** [comment]
Is there anything I may be missing using the latest branch? Located in Canada: @meunierd 
youtube-dl --verbose -f bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo[ext=mkv]+bestaudio[ext=m4a]/bestvideo+bestaudio --no-geo-bypass https://gem.cbc.ca/media/canada-the-story-of-us/season-1/episode-1/38e815a-00beb574afc -u "myemail" -p mypass --rm-cache-dir
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'-f', u'bestvideo[ext=mp4]+bestaudio[ext=m4a]/bestvideo[ext=mkv]+bestaudio[ext=m4a]/bestvideo+bestaudio', u'--no-geo-bypass', u'https://gem.cbc.ca/media/canada-the-story-of-us/season-1/episode-1/38e815a-00beb574afc', u'-u', u'PRIVATE', u'-p', u'PRIVATE', u'--rm-cache-dir']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.01.24
[debug] Python version 2.7.17 (CPython) - Linux-5.3.0-28-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.6, ffprobe 3.4.6
[debug] Proxy map: {}
Removing cache dir /home/user/.cache/youtube-dl ...
[cbc.ca:watch] Acquiring device token
[cbc.ca:watch] 38e815a-00beb574afc: Downloading XML
[cbc.ca:watch] 38e815a-00beb574afc: Downloading XML
[download] Downloading playlist: World's Collide
[cbc.ca:watch] playlist World's Collide: Collected 1 video ids (downloading 1 of them)
[download] Downloading video 1 of 1
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading XML
ERROR: Unable to download XML: HTTP Error 402: Payment Required (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/home/user/.local/lib/python2.7/site-packages/youtube_dl/extractor/common.py", line 627, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/home/bojan/.local/lib/python2.7/site-packages/youtube_dl/YoutubeDL.py", line 2237, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)

**Comment by meunierd** [comment]
I think you might not actually be using my branch judging by where the traceback is. I tried your command locally and it works for me.

**Comment by bojangles89** [comment]
> I think you might not actually be using my branch judging by where the traceback is. I tried your command locally and it works for me.

That was it, thank you!

**Comment by ilike2burnthing** [comment]
Had to clear .cache, but update is working fine!

Hope to see this in the master soon :)

**Comment by tv21** [comment]
Also worked for me after using the --rm-cache-dir option. Would much appreciate this being put in the master version.

**Comment by ilike2burnthing** [comment]
@dstftw is there anything missing here to prevent merging?

Thanks! 👍

**Comment by dstftw** [comment]
Provide working CA proxy.

**Comment by ilike2burnthing** [comment]
Working fine here. Cleared .cache just to make sure. Was able to download without issue.

**Comment by meunierd** [comment]
@dstftw for a CA proxy you can use Tor with these two lines in your torrc:
```
ExitNodes {ca}
StrictNodes 1
```

Restart tor and then pass `--proxy=socks5://localhost:9050` to `youtube-dl`.

**Comment by dstftw** [comment]
I'm well aware of that. All tor end points are banned.

**Comment by meunierd** [comment]
I tested it before commenting and it works:

```
youtube-dl --rm-cache-dir -n --proxy=socks5://localhost:9050 'https://gem.cbc.ca/season/canada-the-story-of-us/season-1/1c9a9fb9-d1d4-4587-9f38-8c1f43d2aee0'
Removing cache dir /Users/meunierd/.cache/youtube-dl ...
[cbc.ca:watch] Downloading JSON metadata
[cbc.ca:watch] Downloading JSON metadata
[cbc.ca:watch] Acquiring device token
[cbc.ca:watch] Downloading XML
[cbc.ca:watch] 1c9a9fb9-d1d4-4587-9f38-8c1f43d2aee0: Downloading XML
[cbc.ca:watch] 1c9a9fb9-d1d4-4587-9f38-8c1f43d2aee0: Downloading XML
[download] Downloading playlist: Season 1
[cbc.ca:watch] playlist Season 1: Collected 10 video ids (downloading 10 of them)
[download] Downloading video 1 of 10
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading XML
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading XML
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading m3u8 information
[download] Destination: World's Collide-38e815a-00c042d36bc.f8687.mp4
WARNING: ffmpeg does not support SOCKS proxies. Downloading is likely to fail. Consider adding --hls-prefer-native to your command.
ffmpeg version 4.2.2 Copyright (c) 2000-2019 the FFmpeg developers
  built with Apple clang version 11.0.0 (clang-1100.0.33.17)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/4.2.2_2 --enable-shared --enable-pthreads --enable-version3 --enable-avresample --cc=clang --host-cflags= --host-ldflags= --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libbluray --enable-libmp3lame --enable-libopus --enable-librubberband --enable-libsnappy --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-librtmp --enable-libspeex --enable-libsoxr --enable-videotoolbox --disable-libjack --disable-indev=jack
```

**Comment by meunierd** [comment]
That said, I do pay for my own VPN that I could privately share. Could we exchange credentials via email?

**Comment by meunierd** [comment]
Testing with the `--hls-prefer-native` flag seems to respect the socks proxy: 

```youtube-dl --rm-cache-dir -n --hls-prefer-native --proxy=socks5://localhost:9050 'https://gem.cbc.ca/season/canada-the-story-of-us/season-1/1c9a9fb9-d1d4-4587-9f38-8c1f43d2aee0'
Removing cache dir /Users/meunierd/.cache/youtube-dl ...
[cbc.ca:watch] Downloading JSON metadata
[cbc.ca:watch] Downloading JSON metadata
[cbc.ca:watch] Acquiring device token
[cbc.ca:watch] Downloading XML
[cbc.ca:watch] 1c9a9fb9-d1d4-4587-9f38-8c1f43d2aee0: Downloading XML
[cbc.ca:watch] 1c9a9fb9-d1d4-4587-9f38-8c1f43d2aee0: Downloading XML
[download] Downloading playlist: Season 1
[cbc.ca:watch] playlist Season 1: Collected 10 video ids (downloading 10 of them)
[download] Downloading video 1 of 10
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading XML
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading XML
[cbc.ca:watch:video] e9923698-08b5-4209-abb3-e9fd76f4d190: Downloading m3u8 information
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 295
[download] Destination: World's Collide-38e815a-00c042d36bc.f8687.mp4
[download]   1.3% of ~2.69GiB at 575.61KiB/s ETA 01:35:00```

**Comment by dstftw** [comment]
As already mentioned this does not work on web needed to verify these changed:
image
Also FYI geo restriction on API calls is already bypassed by using `X-Forwarded-For`.
My email is in profile.

**Comment by meunierd** [comment]
Okay I've emailed the proxy configs. Let me know if there's any trouble with them.

**Comment by meunierd** [comment]
I see that while the `youtube-dl` command succeeds, they ban basically all VPNs at the web level. They're likely using maxmind to catch anything other than residential proxies which I don't have access to. Would you be willing to tunnel through my home internet via SSH?

**Comment by dstftw** [comment]
SSH tunnel is more preferable.

**Comment by ilike2burnthing** [comment]
I'm using this outside Canada with no VPN or manual proxy, and -v shows:

```
[debug] Using fake IP 99.230.181.112 (CA) as X-Forwarded-For.
[cbc.ca:watch] 7d29c147-bb7e-45ac-9263-c47cab0cbe26: Downloading XML
[cbc.ca:watch] 7d29c147-bb7e-45ac-9263-c47cab0cbe26: Downloading XML
[download] Downloading playlist: Season X
[cbc.ca:watch] playlist Season X: Collected XX video ids (downloading XX of them)
[download] Downloading video 1 of XX
[debug] Using fake IP 99.248.81.57 (CA) as X-Forwarded-For.
```

So `X-Forwarded-For` seems to be doing its job fine. Apologies if I'm missing something else.

**Comment by meunierd** [comment]
Could you send me your public key and I'll get it set up.

**Comment by dstftw** [comment]
It's in ytdl downloads page.

**Comment by meunierd** [comment]
I've emailed you the SSH credentials. I tunneled through it locally myself so hopefully everything's okay.

File: youtube_dl/extractor/cbc.py [hunk 4]
```diff
@@ -249,13 +275,19 @@ def _valid_device_token(self):
...
 
     def _register_device(self):
+        signature = self._signature()
         self._device_id = self._device_token = None
         result = self._download_xml(
             self._API_BASE_URL + 'device/register',
             None, 'Acquiring device token',
             data=b'<device><type>web</type></device>')
         self._device_id = xpath_text(result, 'deviceId', fatal=True)
         self._device_token = xpath_text(result, 'deviceToken', fatal=True)
```
Comment (dstftw) [inline youtube_dl/extractor/cbc.py]: This should be under `if signature:` else branch now I suppose.

File: youtube_dl/extractor/cbc.py [hunk 5]
```diff
@@ -249,13 +275,19 @@ def _valid_device_token(self):
...
+        if signature:
+            data = '<login><token>{0}</token><device><deviceId>{1}</deviceId><type>web</type></device></login>'.format(signature, self._device_id).encode()
```
Comment (dstftw) [inline youtube_dl/extractor/cbc.py]: Are you sure `signature`, `self._device_id` will not have `<`, `>`? Strictly speaking those should be escaped or better standard XML facilities should be used for XML building.

**Comment by meunierd** [comment]
I've more clearly separated the signature/authenticated vs anonymous device registration, and escaped the signature using `xml.sax.saxutils.escape` from the standard library.

**Comment by meunierd** [comment]
> There is also a conceptual problem with cache invalidation when you try to authenticate having already cached id and token.
Should I cache which kind of token is in cache? ie if I pass in user credentials, should it only load a signed token from cache?

File: youtube_dl/extractor/cbc.py [hunk 6]
```diff
@@ -249,13 +273,21 @@ def _valid_device_token(self):
...
             self._API_BASE_URL + 'device/register',
             None, 'Acquiring device token',
             data=b'<device><type>web</type></device>')
         self._device_id = xpath_text(result, 'deviceId', fatal=True)
-        self._device_token = xpath_text(result, 'deviceToken', fatal=True)
+        anon_device_token = xpath_text(result, 'deviceToken', fatal=True)
+        email, password = self._get_login_info()
+        if email and password:
+            signature = self._signature(email, password)
+            data = '<login><token>{0}</token><device><deviceId>{1}</deviceId><type>web</type></device></login>'.format(escape(signature), self._device_id).encode()
```
Comment (dstftw) [inline youtube_dl/extractor/cbc.py]: Device id is not escaped.

**Comment by dstftw** [comment]
Caching token kind is not enough since one having already registered with some credentials may want to register under different credentials. So `id` and `token` should be cached per some checksum of particular credentials.

**Comment by makeworld-the-better-one** [comment]
Does this mean main youtube-dl has CBC login support now?

**Comment by ilike2burnthing** [comment]
> Does this mean main youtube-dl has CBC login support now?

Next release

**Comment by ilike2burnthing** [comment]
https://github.com/ytdl-org/youtube-dl/releases/tag/2020.03.24

**Pull request closed without merging**